security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+	// the client to prove server's identities. The rules for setting these two fields are:
+	// Either Certificates or GetCertificate must be set; the other will be ignored
+	//----------------------------------------------------------------------------------------------
+	// If field Certificates is set, field GetClientCertificate will be ignored. The server will use
+	// Certificates every time when asked for a certificate, without performing certificate reloading.
+	Certificates []tls.Certificate
+	// If GetClientCertificate is set and Certificates is nil, the server will invoke this
+	// function every time asked to present certificates to the client when a new connection is
+	// established. This is known as peer certificate reloading.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	// ---------------General Rules For Root Certificate Setting------------------------------------
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the server side. The rules
+	// for setting these two fields are:
+	// If requiring mutual authentication on server side:
+	//     Either RootCACerts or GetRootCAs must be set; the other will be ignored
+	// Otherwise:
+	//     Nothing needed(the two fields will be ignored)
+	//----------------------------------------------------------------------------------------------
+	// If field RootCACerts is set, field GetRootCAs will be ignored. The server will use RootCACerts
+	// every time when verifying the peer certificates, without performing root certificate reloading.
+	RootCACerts *x509.CertPool
+	// If GetRootCAs is set and RootCACerts is nil, the server will invoke this function every time
+	// asked to check certificates sent from the client when a new connection is established.
+	// This is known as root CA certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	// If the server want the client to send certificates to prove its identity.
+	MutualAuth bool
+}
+
+func (o *ClientOptions) config() (*tls.Config, error) {
+	config := &tls.Config{
+		ServerName:           o.ServerNameOverride,
+		Certificates:         o.Certificates,
+		GetClientCertificate: o.GetClientCertificate,
+		RootCAs:              o.RootCACerts,
+		InsecureSkipVerify:   true,
+	}
+	return config, nil
+}
+
+func (o *ServerOptions) config() (*tls.Config, error) {
+	if o.Certificates == nil && o.GetCertificate == nil {
+		return nil, fmt.Errorf("either Certificates or GetCertificate must be specified")
+	}
+	if !o.MutualAuth {
+		return &tls.Config{
+			ClientAuth:     tls.NoClientCert,
+			Certificates:   o.Certificates,
+			GetCertificate: o.GetCertificate,
+		}, nil
+	}
+	if o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs to provide root CA certs if using mutual TLS")
+	}
+	// We fall back to normal config settings if users don't need to reload root certificates.
```

Comment:
We does root credential reloading by rewriting the verification function, and disabling the default check.  We have to disable the default check because otherwise, every time it will use the same static certificate, making reloading impossible. 
On server side, to disable the default check, we have to deliberately set ClientAuth to "NoClientCert"(btw, the equivalent part for client side is to set InSecureSkipVerify to true), which will not do any default check, and then perform custom reloading in the following VerifyPeerCertificate function. 
We have contacted the Go team and this is their suggested way to do root credential reloading. It doesn't look clean enough, and they don't have any plans to change their APIs. But on the other hand, this also proves the utility libraries are beneficial to some users since this might be tricky :-)

Reply:
If it's still relevant, it might be worth adding a comment to summarize this :)